### PR TITLE
[Feature] Concurrent connections & transactions

### PR DESCRIPTION
**Pull request by stevensJourney**
# Description

This PR adds the ability to concurrently access a SQLite DB. The work is based off this [Blog Post](https://www.powersync.co/blog/sqlite-optimizations-for-ultra-high-performance) and adapted from https://github.com/journeyapps/sqlite_async.dart. 

Multiple read connections and a single write connection are opened in WAL mode. A C++ connection pool wardens which context has access to any of the DB connections at a point in time. Asynchronous queuing is handled via the JavaScript side of JSI bindings. All SQL operations are asynchronously executed in the thread pool. 

Single connection DB connections are possible by setting the number of read connections to zero.

## Flow
 The general flow is:
- JavaScript requests read/write lock on a connection via  readLock/writeLock. JS provides a unique context ID for the lock
     
  * If a connection is available: The SQLite connection is locked to the connection lock context ID provided by the requestor. The JavaScript  bridge is informed from the Connection pool that the requested  context lock ID is now active and SQL requests can be made with the context ID (on the relevant connection).
   * If no connections are available, the context ID is added to a FIFO queue. Once other requests are completed the JavaScript bridge is informed that the requested context lock ID is now active.

- Any SQL requests are triggered (at the correct time) from the JavaScript callback. Those requests are synchronized by returning JSI promises over the bridge.

- The JavaScript bridge makes a request to release the lock on the connection pool once it's async callback operations are either resolved or rejected.

## Tests
The test app was updated and now includes tests for DB connections. All tests currently pass on Android and iOS. An automated test for Android has been added to CI.
image

### Related PRs
https://github.com/journeyapps/powersync-react-native-sdk/pull/7 Updated logic to use new DBAdapter API. Tested in demo application.

### CODE REVIEW

File: cpp/ConnectionPool.cpp [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+  struct ConnectionState writeCon;
+  writeCon.connection = nullptr;
+  writeCon.currentLockId = EMPTY_LOCK_ID;
+
+  writeConnection = writeCon;
+
+  // Open the write connection
+  genericSqliteOpenDb(dbName, docPath, &writeConnection.connection,
+                      SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE |
+                          SQLITE_OPEN_NOMUTEX);
```
Comment (rkistner) [inline cpp/ConnectionPool.cpp:25]: Since each connection is used from different threads (even if it's not at the same time in most cases), `SQLITE_OPEN_FULLMUTEX` would be safer.

File: cpp/ConnectionPool.cpp [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+  };
+}
+
+SequelBatchOperationResult
+ConnectionPool::importSQLFile(std::string fileLocation) {
+  std::string line;
+  std::ifstream sqFile(fileLocation);
+
+  if (sqFile.is_open()) {
+    sqlite3 *connection = writeConnection.connection;
```
Comment (rkistner) [inline cpp/ConnectionPool.cpp:258]: This bypasses the locking logic - could result in concurrency issues.

File: cpp/ConnectionPool.cpp [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+   * There is no need to check if mainDBName is opened because
+   * sqliteExecuteLiteral will do that.
+   * */
+  string dbPath = get_db_path(dbFileName, docPath);
+  string statement = "ATTACH DATABASE '" + dbPath + "' AS " + alias;
+
+  auto dbConnections = getAllConnections();
+
+  for (auto &connection : dbConnections) {
+    SequelLiteralUpdateResult result =
```
Comment (rkistner) [inline cpp/ConnectionPool.cpp:210]: This bypasses the locking logic - could result in concurrency issues.

File: src/types.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  loadFileAsync: (dbName: string, location: string) => Promise<FileLoadResult>;
+}
+
+export interface LockOptions {
+  timeoutMs?: number;
+}
+
+export interface LockContext {
+  execute: (sql: string, args?: any[]) => QueryResult;
+  executeAsync: (sql: string, args?: any[]) => Promise<QueryResult>;
```
Comment (rkistner) [inline src/types.ts:127]: It may be better to just have a single asynchronous `execute` method here, same as on the connection.

**Review by rkistner — COMMENTED** [review]
There are currently still possibilities for concurrency issues.

Taking this example:

```ts
await db.writeTransaction(async (tx) => {
   var promises = [];
   for (var i = 0; i < 100; i++) {
      promises.push(tx.executeAsync('SELECT * FROM User'));
   }
   await Promise.all(promises);
});
```

While this transaction has an exclusive lock on the write connection, the individual statements are concurrently executed on different threads, which can cause issues (not sure exactly what issues, but could be anything from a crash to data corruption). Using `SQLITE_OPEN_FULLMUTEX` may mitigate those, but then you'd effectively use all the threads for a single transaction in this case, leaving none available for other concurrent transactions.

A good option would be to combine the ThreadPool and ConnectionPool, and have a dedicated thread and work queue per connection - never letting other threads touch the connection.

File: cpp/ConnectionPool.cpp (hunk 3 above)
Comment (stevensJourney) [reply cpp/ConnectionPool.cpp:210]: This currently operates on all the connections. Adding lock protections to await for when all the connections are available would be tricky. This would most likely need to be executed before any locks are requested. We could add a check to ensure that there are no active locks on any of the connections and fail if there are. Do you think that would be sufficient?

Comment (rkistner) [reply cpp/ConnectionPool.cpp:210]: With the attach call being synchronous, that should be sufficient. I'd assume that in most cases this would be called before any other transactions.

If it's changed to an asynchronous call, we'd need to revisit.

File: cpp/ConnectionPool.cpp (hunk 2 above)
Comment (stevensJourney) [reply cpp/ConnectionPool.cpp:258]: This method is only intended to be used via the connection wrapped in a `writeLock` https://github.com/journeyapps/react-native-quick-sqlite/blob/fd08cef2dac9fa37f97859a2cff9611544ac9e66/src/setup-open.ts#L222. While it may be possible to call the method via the global module, that should not be recommend as using the module directly can bypass certain JS checks in some circumstances.

File: cpp/ConnectionPool.cpp (hunk 3 above)
Comment (stevensJourney) [reply cpp/ConnectionPool.cpp:220]: Added a check to ensure all connections are free.

File: tests/tests/sqlite/rawQueries.spec.ts [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+      expect(res.rows?._array).to.eql([]);
+    });
+
+    it('Correctly throws', async () => {
+      const id = chance.string();
+      const { name, age, networth } = generateUserInfo();
+      try {
+        await db.execute('INSERT INTO "User" (id, name, age, networth) VALUES(?, ?, ?, ?)', [id, name, age, networth]);
+      } catch (e: any) {
+        expect(!!e).to.equal(true);
```
[the remark this answers is not in the document]
Comment (stevensJourney) [reply tests/tests/sqlite/rawQueries.spec.ts]: The test should always throw an error, due to the `id` field being a string and not a number. But you are correct, it would not cover that use case - this test was also a duplicate of `Failed insert`. I removed this tests and updated `Failed insert` to handle the case of if an error is not thrown.

File: cpp/ConnectionPool.cpp (hunk 1 above)
Comment (stevensJourney) [reply cpp/ConnectionPool.cpp]: The connections now open with `SQLITE_OPEN_FULLMUTEX` all the tests still pass correctly. I'm not actually sure why it was using `SQLITE_OPEN_NOMUTEX`.

This module did originally open connections with `SQLITE_OPEN_FULLMUTEX` and it's asynchronous nature allowed different threads to access the connections (with no locks for preventing simultaneous asynchronous operations. I assume if a conflict occurred some operations would fail and throw and error).

File: src/types.ts (hunk 4 above)
Comment (stevensJourney) [reply src/types.ts:127]: 👌 do you think the transaction contexts should only have asynchronous `commit` and `rollback` operations as well?

Comment (rkistner) [reply src/types.ts:127]: Yes, I think it's best if it's all asynchronous. That leaves the option open to change the implementation later, and then same API could also be implemented on other platforms that can only do asynchronous calls.

Comment (stevensJourney) [reply src/types.ts]: 👍 The lock and transaction methods are now all async

**Comment by stevensJourney** [comment]
> There are currently still possibilities for concurrency issues.
> 
> Taking this example:
> 
> ```ts
> await db.writeTransaction(async (tx) => {
>    var promises = [];
>    for (var i = 0; i < 100; i++) {
>       promises.push(tx.executeAsync('SELECT * FROM User'));
>    }
>    await Promise.all(promises);
> });
> ```
> 
> While this transaction has an exclusive lock on the write connection, the individual statements are concurrently executed on different threads, which can cause issues (not sure exactly what issues, but could be anything from a crash to data corruption). Using `SQLITE_OPEN_FULLMUTEX` may mitigate those, but then you'd effectively use all the threads for a single transaction in this case, leaving none available for other concurrent transactions.
> 
> A good option would be to combine the ThreadPool and ConnectionPool, and have a dedicated thread and work queue per connection - never letting other threads touch the connection.

@rkistner The threadpool is now removed and each connection (inside the connection pool) has it's own thread that will queue operations inside the context lock. This will ensure only one DB operation occurs on a connection at a time. A test was also added for this use case.

**Review by rkistner — APPROVED** [review]
Looks great!
